[lldb] Adds additional fields to ProcessInfo

feg208 opened the pull request:
To implement SaveCore for elf binaries we need to populate some additional fields in the prpsinfo struct. Those fields are the nice value of the process whose core is to be taken as well as a boolean flag indicating whether or not that process is a zombie. This commit adds those as well as tests to ensure that the values are consistent with expectations

feg208 committed:
[lldb] Adds additional fields to ProcessInfo

To implement SaveCore for elf binaries we need to populate some
additional fields in the prpsinfo struct. Those fields are the nice
value of the process whose core is to be taken as well as a boolean
flag indicating whether or not that process is a zombie. This commit
adds those as well as tests to ensure that the values are consistent
with expectations

bulbazord commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -237,6 +250,16 @@ class ProcessInstanceInfo : public ProcessInfo {
            m_cumulative_system_time.tv_usec > 0;
   }
 
+  int8_t GetNiceValue() const { return m_nice_value; }
```
Suggestion: `nice` -> `priority_level` or something similar. The unix world uses the words `nice` and `niceness` but Windows and other platforms do not. It's also not a very obvious word if you lack the domain expertise.

emaste replied:
nice and priority are not synonymous though, nice is one input into the process's priority. I'm not sure which we actually want in this context.

feg208 replied:
Done. Agreed. I also added a note in the linux specific code. In linuxland priority implies a realtime priority while nice is nice

feg208 replied:
prpsinfo has pr_nice field to be populated https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/Process/elf-core/ThreadElfCore.h#L99

bulbazord commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -237,6 +250,16 @@ class ProcessInstanceInfo : public ProcessInfo {
+
+  bool NiceValueIsValid() const;
+
+  void SetIsZombie() { m_zombie = true; }
```
Is there a way to unset it being a zombie?

clayborg commented on lldb/source/Utility/ProcessInfo.cpp:
```diff
@@ -17,6 +17,7 @@
 
 #include <climits>
 #include <optional>
+#include <sys/resource.h>
```
Not every OS will have this header file, so I am not sure we can/should be importing this header. Windows build will most likely fail and any non unix based OS as well.

clayborg commented on lldb/source/Utility/ProcessInfo.cpp:
```diff
@@ -117,6 +118,10 @@ bool ProcessInfo::IsScriptedProcess() const {
 }
 
+bool ProcessInstanceInfo::NiceValueIsValid() const {
+  return m_nice_value >= PRIO_MIN && m_nice_value <= PRIO_MAX;
```
`PRIO_MIN` and `PRIO_MAX` come from linux/unix header and won't be available on other systems. I am also wondering if the `PRIO_MIN` and `PRIO_MAX` will differ from one linux OS to another? Probably best to not rely on these defines from OS specific header files, so I would suggest either manually defining them in this file or not checking the values.

feg208 replied:
I removed this and just made the check to be a value stored in the optional

clayborg commented on lldb/unittests/Host/linux/HostTest.cpp:
```diff
@@ -12,6 +12,9 @@
 #include "gtest/gtest.h"
 
+#include <cerrno>
+#include <sys/resource.h>
```
it is ok to include this header file here because this is a linux specific host test.

feg208 replied:
yeah. The cmake file enforces a linux-only build as well. However as @jimingham pointed out someone could plausibly move this into a non-linux build. Maybe I should place guards here and below? The functionality is in POSIX I think and my understanding is that windows would conform to POSIX to some degree. But to be honest I know very little about what happens on the windows side of the fence.

emaste replied:
Much of this file would work just fine on FreeBSD as well so that might make sense, although I'm not sure what the best structure would be -- maybe much of this file should move to a new lldb/unittests/Host/posix/HostTest.cpp, leaving anything Linux-specific here?

clayborg replied:
This is fine if this is a host test, so nothing needs to be done. We are expecting linux + posix here which is fine.

clayborg reviewed (changes requested):
So `ProcessInstanceInfo` is already very OS specific, so I can't object to adding more OS specific information to this structure. It would be nice to abstract OS specific things into a dictionary so allow OS specific things to be added with OS specific key/value pairs, but since it is already very unix centric these changes are ok. We have to be careful to not include OS specific header files like `<sys/resource.h>` as they won't be available on windows or other non linux OS versions. See inline comments for details.

feg208 commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -144,6 +144,19 @@ class ProcessInstanceInfo : public ProcessInfo {
     long int tv_usec = 0;
   };
 
+  enum class ProcessState {
+    Unknown,
+    Dead,
+    DiskSleep,
+    Idle,
+    Paging,
+    Parked,
+    Running,
+    Sleeping,
+    TracedOrStopped,
+    Zombie,
+  };
+
```
yeah. So if you look at the definition of ELFLinuxPrPsInfo there is a char member `pr_state` that will be something like T or Z or whatever. So we'd need these available to set that assuming we didn't want OS specific codes in there which are very linux specific (I found the mapping in the linux kernel code). Maybe we should move it back though? I mean for the purpose of generating a core the state is always going to be TraceOrStopped in any case.

clayborg replied:
There doesn't seem to be a reason for this to be here, so if we can move it back, that would be great. Ideally this file would be OS agnostic, but it started out a bit unix specific in the first place.

feg208 replied:
moved back

feg208 commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -254,6 +277,8 @@ class ProcessInstanceInfo : public ProcessInfo {
   struct timespec m_system_time {};
   struct timespec m_cumulative_user_time {};
   struct timespec m_cumulative_system_time {};
+  int8_t m_nice_value = INT8_MAX;
```
yeah. I'll make this change. Especially here since priority can be in a range that has zero near the center

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -254,6 +280,8 @@ class ProcessInstanceInfo : public ProcessInfo {
   struct timespec m_cumulative_user_time {};
   struct timespec m_cumulative_system_time {};
+  std::optional<int8_t> m_priority_value = std::nullopt;
+  bool m_zombie = false;
```
making this optional as well would be good. If it isn't set, it means we don't know, but if it is set to true/false we do know. Again, almost all members of this structure could benefit from being std::optional

feg208 replied:
done

clayborg commented on lldb/source/Host/linux/Host.cpp:
```diff
@@ -91,14 +87,16 @@ static bool GetStatusInfo(::pid_t Pid, ProcessInstanceInfo &ProcessInfo,
+          &stat_fields.minflt, &stat_fields.cminflt, &stat_fields.majflt,
+          &stat_fields.cmajflt, &stat_fields.utime, &stat_fields.stime,
+          &stat_fields.cutime, &stat_fields.cstime,
+          &stat_fields.realtime_priority, &stat_fields.priority) < 0) {
```
are the priority fields always in this data? Just want to make sure that this `sscanf` call won't fail on some certain versions of linux/unix?

feg208 replied:
from proc_pid_stat(5)
```
(18) priority  %ld
                     (Explanation for Linux 2.6) For processes running a real-time scheduling policy (policy below; see sched_setscheduler(2)), this is the negated scheduling priority, minus one; that is, a number in the range -2 to -100, corresponding to  real-time
                     priorities  1  to  99.   For processes running under a non-real-time scheduling policy, this is the raw nice value (setpriority(2)) as represented in the kernel.  The kernel stores nice values as numbers in the range 0 (high) to 39 (low), corre‐
                     sponding to the user-visible nice range of -20 to 19.

                     Before Linux 2.6, this was a scaled value based on the scheduler weighting given to this process.
```
so yes

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -147,96 +148,111 @@ class ProcessInstanceInfo : public ProcessInfo {
   }
 
-  uint32_t GetEffectiveUserID() const { return m_euid; }
+  uint32_t GetEffectiveUserID() const { return m_euid.value(); }
```
can we actually return `std::optional<uint32_t>` here? Otherwise if someone calls this and `m_euid` doesn't have a value, then it will throw a `std::bad_optional_access` exception.

There is a `std::optional<T>::value_or(T)`, but that will defeat the reason for adding a std::optional in the first place. So I see two options:
```
uint32_t GetEffectiveUserID(uint32_t default_value = 0) const { return m_euid.value_or(default_value); }
```
or
```
std::optional<uint32_t> GetEffectiveUserID() const { return m_euid; }
```
I would prefer the latter, but now sure now many locations this will affect.

feg208 replied:
I don't have particularly strong feelings here but we should keep it consistent between this class and the parent class which has GetUID and GetGID neither of which returns a std::optional. Given the risk of std::bad_optional_access getting thrown thoughout I am coming to the notion that maybe we shouldn't use optionals here? If we have optionals I guess we are mandating that the paired IsValid must exist and must always be called. That seems like it'll create challenges. Can we go back to using invalid values like UINT32_MAX and maybe alter the interface to use optionals in a subsequent pr? This way we could debate the merits of this approach independent of adding a niceness field.

feg208 replied:
Maybe a better approach is to keep the optionals in the members that are particular to ProcessInstanceInfo (i.e. the time fields, priority, and the zombie fields) and revert the pid fields and the pid and uid/gid fields to the previous and punt that to a later pr. How does that sound?

feg208 replied:
@clayborg I made the changes referenced above (i.e. backed out the optionals for the pid/uid/gid fields). Is this a dealbreaker for you? Or do you think we can revisit at a later time?

clayborg replied:
It is ok to not do all of the ivars at one, but for the `std::optional` stuff that remains, if we just call `.value()` and it hasn't been set we are going to throw an exception right? We either need to make these not optional, or return the optional when accessing and let the callers deal with the values no being available.

feg208 replied:
yeah. I of the same view

hawkinsw commented on lldb/source/Host/linux/Host.cpp:
```diff
@@ -70,6 +71,12 @@ struct StatFields {
   long unsigned stime;
   long cutime;
   long cstime;
+  // in proc_pid_stat(5) this field is specified as priority
```
```suggestion
  // In proc_pid_stat(5) this field is specified as priority
```

feg208 replied:
done

hawkinsw commented on lldb/source/Host/linux/Host.cpp:
```diff
@@ -70,6 +71,12 @@ struct StatFields {
   long cstime;
+  // in proc_pid_stat(5) this field is specified as priority
+  // but documented as realtime priority. To keep with the adopted
+  // nomenclature in ProcessInstanceInfo we adopt the documented
```
```suggestion
  // nomenclature in ProcessInstanceInfo, we adopt the documented
```

feg208 replied:
done

hawkinsw reviewed (commented):
Sorry for the nits! I am not smart enough to contribute on the technical merits, so I am trying to find some way to help!

feg208 commented on lldb/unittests/Host/linux/HostTest.cpp:
```diff
@@ -86,4 +89,22 @@ TEST_F(HostTest, GetProcessInfo) {
+  int prio = getpriority(PRIO_PROCESS, PRIO_PROCESS);
+  ASSERT_TRUE((prio < 0 && errno == 0) || prio >= 0);
+  ASSERT_EQ(Info.GetPriorityValue(), prio);
+  // If we can't raise our nice level then this test can't be performed
```
done

feg208 commented on lldb/source/Host/linux/Host.cpp:
```diff
@@ -115,13 +124,19 @@ static bool GetStatusInfo(::pid_t Pid, ProcessInstanceInfo &ProcessInfo,
   };
 
+  // priority (nice) values run from 19 to -20 inclusive (in linux). In the
+  // prpsinfo struct pr_nice is a char
```
done

feg208 commented on lldb/source/Host/linux/Host.cpp:
```diff
@@ -115,13 +124,19 @@ static bool GetStatusInfo(::pid_t Pid, ProcessInstanceInfo &ProcessInfo,
     return ts;
   };
 
+  // priority (nice) values run from 19 to -20 inclusive (in linux). In the
```
done

feg208 commented on lldb/source/Host/linux/Host.cpp:
```diff
@@ -70,6 +71,12 @@ struct StatFields {
+  // in proc_pid_stat(5) this field is specified as priority
+  // but documented as realtime priority. To keep with the adopted
+  // nomenclature in ProcessInstanceInfo we adopt the documented
+  // naming here
```
done

feg208 commented:
> Sorry for the nits! I am not smart enough to contribute on the technical merits, so I am trying to find some way to help!

Ehhh grammar is not where I shine. I rolled them up

hawkinsw commented:
> > Sorry for the nits! I am not smart enough to contribute on the technical merits, so I am trying to find some way to help!
> 
> Ehhh grammar is not where I shine. I rolled them up

Not at all! I thought the comments were really excellent and helpful! I just wanted to help if I could!

feg208 commented:
Hello all. I am wondering what I can do to advance this patch? I think it is required to support process save-core in linux for lldb. I'd love to move this before adding static methods in ThreadEfCore.h to produce populated prpsinfo and prstatus structs for inclusion in a generated corefile. Is there someone else I should include on review here? @JDevlieghere? I am happy to amend this in whatever way the community thinks works best for lldb.

JDevlieghere commented:
> Hello all. I am wondering what I can do to advance this patch? I think it is required to support process save-core in linux for lldb. I'd love to move this before adding static methods in ThreadEfCore.h to produce populated prpsinfo and prstatus structs for inclusion in a generated corefile. Is there someone else I should include on review here? @JDevlieghere? I am happy to amend this in whatever way the community thinks works best for lldb.

A message like this that the PR is ready for re-review should be all you need. Several folks left comments and they might be waiting for all of them to be addressed before they take another look. If you don't hear back within a week, you can ping the patch to get reviewer's attention. 

There's some more info here: https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted

feg208 commented:
> A message like this that the PR is ready for re-review should be all you need. Several folks left comments and they might be waiting for all of them to be addressed before they take another look. 

Ok. @clayborg I think I have addressed your concerns but I'd love another look as you have time.

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -250,10 +268,12 @@ class ProcessInstanceInfo : public ProcessInfo {
-  struct timespec m_system_time {};
-  struct timespec m_cumulative_user_time {};
-  struct timespec m_cumulative_system_time {};
+  std::optional<struct timespec> m_user_time = std::nullopt;
+  std::optional<struct timespec> m_system_time = std::nullopt;
+  std::optional<struct timespec> m_cumulative_user_time = std::nullopt;
+  std::optional<struct timespec> m_cumulative_system_time = std::nullopt;
```
Revert all `struct timespec` optionals because we aren't checking if they have values and we can crash when calling the accessors.

feg208 replied:
done

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -250,10 +268,12 @@ class ProcessInstanceInfo : public ProcessInfo {
+  std::optional<struct timespec> m_system_time = std::nullopt;
+  std::optional<struct timespec> m_cumulative_user_time = std::nullopt;
+  std::optional<struct timespec> m_cumulative_system_time = std::nullopt;
+  std::optional<int8_t> m_priority_value = std::nullopt;
+  std::optional<bool> m_zombie = std::nullopt;
```
If we are going to make these optional, then the accessors should return optionals

feg208 replied:
done

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -195,48 +195,66 @@ class ProcessInstanceInfo : public ProcessInfo {
+
+  bool PriorityValueIsValid() const;
+
+  void SetIsZombie(bool is_zombie) { m_zombie = is_zombie; }
```
The setter here is fine

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -195,48 +195,66 @@ class ProcessInstanceInfo : public ProcessInfo {
+
+  void SetIsZombie(bool is_zombie) { m_zombie = is_zombie; }
+
+  bool IsZombieValid() const { return m_zombie.has_value(); }
+
+  bool IsZombie() const { return m_zombie.value(); }
```
If we are going to use optionals for m_zombie, then we don't need two functions, just return the optional:
```
std::optional<bool> GetIsZombie() const { return m_zombie; }
```
And callers will need to check if this has a value and set to a default value like:
```
bool is_zombie = proc_info.GetIsZombie().value_or(false);
```

feg208 replied:
done

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -195,48 +195,66 @@ class ProcessInstanceInfo : public ProcessInfo {
+            m_cumulative_system_time->tv_usec > 0);
   }
 
+  int8_t GetPriorityValue() const { return m_priority_value.value(); }
```
This function should return `std::optional<int8_t>` if we are going to use an optional value for `m_priority_value` as this is just a crash waiting to happen. Callers can provide figure out a good default value to use if this hasn't been set with code like:
```
int8_t priority = proc_info. GetPriorityValue().value_or(0);
```
But as it stands we are going to crash at some point if someone doesn't call `PriorityValueIsValid()` first. So lets either use std::optional and return std::optional, or just default to sane values without optionals.

feg208 replied:
done

clayborg commented on lldb/include/lldb/Utility/ProcessInfo.h:
```diff
@@ -195,48 +195,66 @@ class ProcessInstanceInfo : public ProcessInfo {
+    m_priority_value = priority_value;
+  }
+
+  bool PriorityValueIsValid() const;
```
This should be removed as this is the point of using std::optionals.

feg208 replied:
done

clayborg reviewed (changes requested):
We should either truly use `std::optional` and return `std::optional<T>` for all getters, or just not use them. This patch adds a `bool IsZombieValid()` which the user must know to call prior to calling `bool IsZombie()`, which is just a crash waiting to happen. So lets either use `std::optional` correclty, or remove it and use sane default values. I like the idea of using `std::optional` and suggested ways to provide default values with the `value_or(T)` calls clients can easily use.

So might suggest to revert any `struct timespec` optionals since we aren't using them at all, and use `std::optional` and return `std::optional` from the getter functions and have the clients use `value_or(T)` just for these two new functions.

feg208 commented:
@clayborg I rolled up those requested changes. That'll make a nicer interface

feg208 commented:
@clayborg This pr is ready for re-review

feg208 commented:
@clayborg This pr is ready for re-review

clayborg commented on lldb/unittests/Host/linux/HostTest.cpp:
```diff
@@ -86,4 +89,22 @@ TEST_F(HostTest, GetProcessInfo) {
+  if (max_incr < prio) {
+    EXPECT_EQ(setpriority(PRIO_PROCESS, PRIO_PROCESS, prio - 1), 0);
+    ASSERT_TRUE(Host::GetProcessInfo(getpid(), Info));
+    ASSERT_EQ(Info.GetPriorityValue().value(), prio - 1);
```
Add an assert to make sure `Info.GetPriorityValue().has_value()` like you did below for the zombie.

feg208 replied:
Done. Thanks for all your effort reviewing this work

clayborg reviewed (approved):
One nit from inlined code to check `Info.GetPriorityValue().has_value()` and this is good to go.

clayborg reviewed (approved):
LGTM

feg208 commented:
@clayborg I don't have commit access. Can you merge this?

feg208 commented:
@JDevlieghere could I talk you into merging this? Somehow I got shifted from the committers group to the triagers group but it's taking time to resolve.

feg208 commented:
@JDevlieghere and @clayborg sorry for the line noise. It's merged
